dfdaemon/config/config.go

```diff
@@ -0,0 +1,149 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"regexp"
+
+	"gopkg.in/yaml.v2"
+)
+
+// -----------------------------------------------------------------------------
+// Properties
+
+// NewProperties create a new properties with default values.
+func NewProperties() *Properties {
+	return &Properties{}
+}
+
+// Properties holds all configurable properties of dfdaemon.
+// The default path is '/etc/dragonfly/dfdaemon.yml'
+type Properties struct {
+	Registries []*Registry `yaml:"registries"`
+}
+
+// Load loads properties from config file.
+func (p *Properties) Load(path string) error {
+	if err := p.loadFromYaml(path); err != nil {
+		return err
+	}
+	var tmp []*Registry
+	for _, v := range p.Registries {
+		if v != nil {
+			if err := v.init(); err != nil {
+				return err
+			}
+			tmp = append(tmp, v)
+		}
+	}
+	p.Registries = tmp
+	return nil
+}
+
+func (p *Properties) loadFromYaml(path string) error {
+	yamlFile, err := ioutil.ReadFile(path)
+	if err != nil {
+		return fmt.Errorf("read yaml config from %s error: %v", path, err)
+	}
+	err = yaml.Unmarshal(yamlFile, p)
+	if err != nil {
+		return fmt.Errorf("unmarshal yaml error:%v", err)
+	}
+	return nil
+}
+
+// -----------------------------------------------------------------------------
+// Registry
+
+// NewRegistry create and init registry proxy with the giving values.
+func NewRegistry(schema, host, cert, regx string) (*Registry, error) {
+	reg := &Registry{
+		Schema: schema,
+		Host:   host,
+		Cert:   cert,
+		Regx:   regx,
+	}
+	if err := reg.init(); err != nil {
+		return nil, err
+	}
+	return reg, nil
+}
+
+// Registry is the proxied registry base information.
+type Registry struct {
+	// Schema can be 'http', 'https' or empty. It will use dfdaemon's schema if
+	// it's empty.
+	Schema string `yaml:"schema"`
+
+	// Host is the host of proxied registry, including ip and port.
+	Host string `yaml:"host"`
+
+	// Cert is the path of server-side certification. It should be provided when
+	// the 'Schema' is 'https' and the dfdaemon is worked on proxy pattern and
+	// the proxied registry is self-certificated.
+	// The server-side certification could be get by using this command:
+	// openssl x509 -in <(openssl s_client -showcerts -servername xxx -connect xxx:443 -prexit 2>/dev/null)
+	Cert string `yaml:"cert"`
+
+	// Regx is a regular expression, dfdaemon use this registry to process the
+	// requests whose host is matched.
+	Regx string `yaml:"regx"`
+
+	compiler  *regexp.Regexp
+	tlsConfig *tls.Config
+}
+
+// Match reports whether the Registry matches the string s.
+func (r *Registry) Match(s string) bool {
+	return r.compiler != nil && r.compiler.MatchString(s)
+}
+
+// TLSConfig returns a initialized tls.Config instance.
+func (r *Registry) TLSConfig() *tls.Config {
+	return r.tlsConfig
+}
+
+func (r *Registry) init() error {
```

Comment:
Shall we have a validation for the parameters，such as "schema". And the test should contain the corresponding test instance.

Reply:
done